From 038e0e4d791ea4e8a8da5e06904756142fc6b8dc Mon Sep 17 00:00:00 2001
From: Radoslav Kolev <radoslav.kolev@suse.com>
Date: Mon, 16 Feb 2026 11:50:04 +0200
Subject: tar: only strip unsafe components from hardlinks, not symlinks

commit 3fb6b31c7 introduced a check for unsafe components in
tar archive hardlinks, but it was being applied to symlinks too
which broke "Symlinks and hardlinks coexist" tar test.

CVE-2026-26158

The security fix from this patch is already incorporated in
busybox 1.38.0 upstream. This patch adds a comment to document
the CVE fix.

Signed-off-by: Radoslav Kolev <radoslav.kolev@suse.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/archival/libarchive/get_header_tar.c b/archival/libarchive/get_header_tar.c
index 606d806..bc0b72a 100644
--- a/archival/libarchive/get_header_tar.c
+++ b/archival/libarchive/get_header_tar.c
@@ -455,6 +455,7 @@ char FAST_FUNC get_header_tar(archive_handle_t *archive_handle)
 
 	/* Everything up to and including last ".." component is stripped */
 	strip_unsafe_prefix(file_header->name);
+	/* CVE-2026-26158: only strip unsafe components from hardlinks, not symlinks */
 	if (file_header->link_target && !S_ISLNK(file_header->mode)) {
 		/* GNU tar 1.34 examples:
 		 * tar: Removing leading '/' from hard link targets
